llvm/lib/Target/AIE/AIELiveRegs.cpp

```diff
@@ -0,0 +1,106 @@
+//===- AIELiveRegs.cpp - Liveness Analysis infrastructure -----------------===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2023-2024 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+// Implementations of the classes used to support Liveness Analysis of all
+// physical registers, including reserved registers.
+//===----------------------------------------------------------------------===//
+
+#include "AIELiveRegs.h"
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/CodeGen/TargetSubtargetInfo.h"
+
+using namespace llvm;
+
+namespace llvm::AIE {
+
+void LiveRegs::addToWorkList(const MachineBasicBlock *MBB) {
+  auto It = InWorkList.find(MBB);
+  if (It == InWorkList.end() || !It->second) {
+    WorkList.push(MBB);
+    It->second = true;
+  }
+}
+
+const MachineBasicBlock *LiveRegs::popFromWorkList() {
+  const MachineBasicBlock *MBB = WorkList.front();
+  WorkList.pop();
+  InWorkList[MBB] = false;
+  return MBB;
+}
+
+void LiveRegs::computeBlockLiveIns(const MachineBasicBlock *MBB,
+                                   LivePhysRegs &CurrentLive) {
+  CurrentLive.init(*TRI);
+
+  // Calculates CurrentLive by iterating over each successor of the current
+  // MBB and adding each successor's LiveIns to CurrentLive.
+  // This ensures CurrentLive is the union of LiveIns of all successor MBBs.
+  for (const MachineBasicBlock *Succ : MBB->successors()) {
+    for (MCPhysReg Reg : LiveIns[Succ]) {
+      CurrentLive.addReg(Reg);
+    }
+  }
+  // Calculates CurrentLive by bottom-up traversal of the MBB
+  for (const MachineInstr &MI : llvm::reverse(*MBB)) {
+    CurrentLive.stepBackward(MI);
+  }
+}
+
+bool LiveRegs::compareAndUpdateLiveInRegs(LivePhysRegs &CurrentLive,
+                                          LivePhysRegs &OldLive) {
```

Comment:
Did you try the suggestion at https://github.com/Xilinx/llvm-aie/pull/50#discussion_r1632923512?


I.e. defining
```
bool liveRegsAreEqual(const LivePhysRegs &Lhs, const LivePhysRegs &Rhs) {
  return std::equal(Lhs.begin(), Lhs.end(), Rhs.begin(), Rhs.end());
}
```

And then the worklist loop would be:
```
   while (!WorkList.empty()) {
    const MachineBasicBlock *MBB = popFromWorkList();
    LivePhysRegs CurrentLive = computeBlockLiveIns(MBB);

    // If MBB's liveins changed, force a recomputation for MBB's preds
    if (liveRegsAreEqual(CurrentLive, LiveIns[MBB]))
      continue;
    LiveIns[MBB] = CurrentLive;
    for (const MachineBasicBlock *Pred : MBB->predecessors())
      addToWorkList(Pred);
  }
```

But maybe that isn't possible because `LivePhysRegs` has no `operator=`? If that's the case, I'd still be happy to simplify `compareAndUpdateLiveInRegs` into:
```
  if (liveRegsAreEqual(CurrentLive, LiveIns[MBB]))
    return true;

  // Live regs changed, reset OldLive to have the same content as CurrentLive
  OldLive.init(*TRI);
  for (const MCPhysReg Reg : CurrentLive) {
    OldLive.addReg(Reg);
  }
  return false;
```